Sources/AWSLambdaRuntimeCore/HTTPClient.swift

```diff
@@ -149,133 +156,75 @@ internal final class HTTPClient {
     }
 }
 
-private final class HTTPHandler: ChannelDuplexHandler {
-    typealias OutboundIn = HTTPClient.Request
-    typealias InboundOut = HTTPClient.Response
-    typealias InboundIn = HTTPClientResponsePart
-    typealias OutboundOut = HTTPClientRequestPart
-
-    private let keepAlive: Bool
-    private var readState: ReadState = .idle
-
-    init(keepAlive: Bool) {
-        self.keepAlive = keepAlive
-    }
-
-    func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
-        let request = unwrapOutboundIn(data)
-
-        var head = HTTPRequestHead(version: .init(major: 1, minor: 1), method: request.method, uri: request.url, headers: request.headers)
-        head.headers.add(name: "host", value: request.targetHost)
-        switch request.method {
-        case .POST, .PUT:
-            head.headers.add(name: "content-length", value: String(request.body?.readableBytes ?? 0))
-        default:
-            break
-        }
-
-        // We don't add a "Connection" header here if we want to keep the connection open,
-        // HTTP/1.1 defines specifies the following in RFC 2616, Section 8.1.2.1:
-        //
-        // An HTTP/1.1 server MAY assume that a HTTP/1.1 client intends to
-        // maintain a persistent connection unless a Connection header including
-        // the connection-token "close" was sent in the request. If the server
-        // chooses to close the connection immediately after sending the
-        // response, it SHOULD send a Connection header including the
-        // connection-token close.
-        //
-        // See also UnaryHandler.channelRead below.
-        if !self.keepAlive {
-            head.headers.add(name: "connection", value: "close")
-        }
-
-        context.write(self.wrapOutboundOut(HTTPClientRequestPart.head(head))).flatMap { _ -> EventLoopFuture<Void> in
-            if let body = request.body {
-                return context.writeAndFlush(self.wrapOutboundOut(HTTPClientRequestPart.body(.byteBuffer(body))))
-            } else {
-                context.flush()
-                return context.eventLoop.makeSucceededFuture(())
-            }
-        }.cascade(to: promise)
-    }
-
-    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
-        let response = unwrapInboundIn(data)
-
-        switch response {
-        case .head(let head):
-            guard case .idle = self.readState else {
-                preconditionFailure("invalid read state \(self.readState)")
-            }
-            self.readState = .head(head)
-        case .body(var bodyPart):
-            switch self.readState {
-            case .head(let head):
-                self.readState = .body(head, bodyPart)
-            case .body(let head, var body):
-                body.writeBuffer(&bodyPart)
-                self.readState = .body(head, body)
```

Comment:
We have a fairly large performance issue here... Since `body`'s internal storage is referenced from `self.readState` and from `body`, we incur a CoW on every append to the local buffer. We could fix this by an easy CoW prevention like:

```swift
case .body(let head, var body):
    self.readState = .idle // removes the self.readState reference. Therefore no Cow necessary
    body.writeBuffer(&bodyPart)
    self.readState = .body(head, body)
```

However since October of last year NIO has a `NIOHTTPClientResponseAggregator`. I think we should use that instead. All the http write logic from `HTTPHandler` has been moved into `UnaryHandler` directly.

Reply:
moving to `NIOHTTPClientResponseAggregator` makes perfect sense. I wonder tho if moving all the extra functionality to `UnaryHandler` is overloading its intent - to be a unary handler. not sure the performance costs of having another handlers, if low, we can consider splitting that up.